### PR TITLE
Mitigate a conflict when using sentencepiece

### PR DESCRIPTION
**Pull request by tengomucho**
# What does this PR do?

This is a fix to a conflict between the fast tokenizers usage and sentencepiece module.
This is due to the fact that protobuf C implementation uses a global pool for all added descriptors, so if two different files add descriptors, they will end up conflicting.

I added a test showing the problem and a guard when loading the proto file to mitigate the problem. Note that the problem is not completly removed: if for any obscure reason an invalid sentencepiece proto file was loaded before, it will keep on using that. Also, if sentencepiece was loaded _after_ the tokenizers load the proto file, the error will occur again (but at least there will be a way to avoid it).

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Original discussion in internal slack: https://huggingface.slack.com/archives/C014N4749J9/p1725371896664959
- [x] Did you write any new necessary tests?

## Who can review?

@ydshieh

### CODE REVIEW

File: src/transformers/utils/sentencepiece_model_pb2_new.py
Comment (ydshieh) [inline src/transformers/utils/sentencepiece_model_pb2_new.py:20]: My concern here is: when we get this, does it (always) works with `transformers`? (we might get strange stuff for this object from other libraries).

File: tests/tokenization/test_tokenization_utils.py
Comment (ydshieh) [inline tests/tokenization/test_tokenization_utils.py:310]: when we do this, we are importing `sentencepiece.sentencepiece_model_pb2` and it will persists druing the whole (remaining) life time of the python test process.

I would probably avoid this, and run this test in a separate test

File: src/transformers/utils/sentencepiece_model_pb2_new.py
Comment (tengomucho) [reply src/transformers/utils/sentencepiece_model_pb2_new.py:20]: No, we do not have a real guarantee that it will _always_ work. That is why I said it mitigates the issues, it doesn't remove it altogether. Having said that, if some other code loaded before transformers uses protobuf and adds a file whose name matches `sentencepiece_model.proto`, I guess it's not going to be anything that is not sentencepiece! Alternatively I could do it the other way around, I can try to add the fail, and if it raises a `TypeError` I can try to do the `FindFileByName`.
In both cases I could also add a warning if `FindFileByName` is used.

File: tests/tokenization/test_tokenization_utils.py
Comment (tengomucho) [reply tests/tokenization/test_tokenization_utils.py:310]: are you suggesting to move this test to its own file?

File: src/transformers/utils/sentencepiece_model_pb2_new.py
Comment (ydshieh) [reply src/transformers/utils/sentencepiece_model_pb2_new.py:20]: No, I don't mean `it's sentencepiece or not sentencepiece`. What I worry a bit is it's (may) not the one we have and we don't know what would happen ...

File: tests/tokenization/test_tokenization_utils.py
Comment (ydshieh) [reply tests/tokenization/test_tokenization_utils.py:310]: sorry, typo, I mean in a separate python subprocess.

But we can wait for a core maintainer's review.

File: src/transformers/utils/sentencepiece_model_pb2_new.py
Comment (ydshieh) [reply src/transformers/utils/sentencepiece_model_pb2_new.py:20]: that could probably be controlled by an  environment variable however.

File: tests/tokenization/test_tokenization_utils.py
Comment (tengomucho) [reply tests/tokenization/test_tokenization_utils.py:310]: OK! Who is the core maintainer? Can we tag him?

File: src/transformers/utils/sentencepiece_model_pb2_new.py
Comment (tengomucho) [reply src/transformers/utils/sentencepiece_model_pb2_new.py:20]: the problem is that is the `sentencepiece_model.proto` was added in the current process, there is no way to add it again. We cannot be sure it's the right one, as it does not provide a version or a hash. And if we try to load it, it will raise an exception from the C extension. So that is why I thought that assuming it's the same proto is probably a fair fallback solution.

Comment (tengomucho) [reply src/transformers/utils/sentencepiece_model_pb2_new.py:20]: if you have a better option, I'd be happy to consider it though!

File: tests/tokenization/test_tokenization_utils.py
Comment (ydshieh) [reply tests/tokenization/test_tokenization_utils.py:310]: I already ping Arthur

**Comment by ArthurZucker** [comment]
Hey! Could you provide the reproducer as well? 
I really don't mind merging as is, but just want to make sure, you 1st have a project that imported protobuf, and 2. used `from transformers.convert_slow_tokenizer import import_protobuf` and that failed right? 
What is a bit weird is we have this guard:

```python
def import_protobuf(error_message=""):
    if is_protobuf_available():
        import google.protobuf

        if version.parse(google.protobuf.__version__) < version.parse("4.0.0"):
            from transformers.utils import sentencepiece_model_pb2
        else:
            from transformers.utils import sentencepiece_model_pb2_new as sentencepiece_model_pb2
        return sentencepiece_model_pb2
    else:
        raise ImportError(PROTOBUF_IMPORT_ERROR.format(error_message))
```
is there no way to check if it was not already imported? (meaning try important, if the error is import error, we expect ?

**Comment by tengomucho** [comment]
Hello! I saw first this problem because I tried to use [Jetstream/Pytorch](https://github.com/google/jetstream-pytorch), and they use sentencepiece and its [`sentencepiece_pb.py`](https://github.com/google/sentencepiece/blob/master/python/src/sentencepiece/sentencepiece_model_pb2.py). In transformers there is [a copy](https://github.com/huggingface/transformers/blob/main/src/transformers/utils/sentencepiece_model_pb2_new.py) of the same file, with few changes to make some variables globals. So, while in python world they are technically two different modules, the C protobuf extension shares the same space, and they conflict when calling `AddSerializedFile`. That's why I proposed this option, because I didn't find any other way to figure out how to avoid this conflict.

I was just thinking now that an alternative could be to check if the `sentencepiece` module is available, import the `sentencepiece_pb.py` from that package, and avoid conflicts. We could then delete our "sentencepiece_pb_new.py" file. But that means it will only work when the `sentencepiece` module is available.

**Comment by tengomucho** [comment]
I force-pushed a cleaner solution, I check if `sentencepiece` is available and if it's the case, we use that instead of our proto version, so we avoid conflicts.

**Review by ArthurZucker — APPROVED** [review]
A lot better now, thanks

**Pull request merged**
